### PR TITLE
Do not run through following phases in Rewriter.LowerBodyOrInitializer once errors are encountered.

### LINKED ISSUE #2774: Assert in CodeGenerator.EmitConversionExpression
```VB
            Const source = "
Imports System.Linq

Class C
    Sub M(args As String())
        Dim concat =
            From x In args
            Let y = x.ToString()
            Let z = x.GetHashCode()
            Select x & y & z
    End Sub
End Class
"

            Dim comp = CreateCompilationWithMscorlib({source}, {SystemCoreRef}, TestOptions.DebugDll)
            comp.VerifyEmitDiagnostics()
```

```
Debug.Assert() Failure
   at System.Diagnostics.TraceListener.Fail(String message)
   at System.Diagnostics.TraceInternal.Fail(String message)
   at System.Diagnostics.Debug.Assert(Boolean condition)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitConversion.vb(189,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitConversionExpression(BoundConversion conversion, Boolean used)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitExpression.vb(58,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitExpression.vb(478,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitArguments(ImmutableArray`1 arguments, ImmutableArray`1 parameters)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitExpression.vb(1007,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitCallExpression(BoundCall call, Boolean used)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitExpression.vb(40,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\CodeGenerator.vb(211,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitSequencePointExpression(BoundSequencePointExpression node, Boolean used)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitExpression.vb(100,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitStatement.vb(868,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitReturnStatement(BoundReturnStatement boundReturnStatement)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitStatement.vb(38,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitStatement(BoundStatement statement)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitStatement.vb(1200,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitBlock(BoundBlock scope)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\EmitStatement.vb(16,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitStatement(BoundStatement statement)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\CodeGenerator.vb(144,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.GenerateImpl()
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\CodeGen\CodeGenerator.vb(88,0): at Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.Generate()
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\Compilation\MethodCompiler.vb(1523,0): at Microsoft.CodeAnalysis.VisualBasic.MethodCompiler.GenerateMethodBody(PEModuleBuilder moduleBuilder, MethodSymbol method, Int32 methodOrdinal, BoundStatement block, ImmutableArray`1 lambdaDebugInfo, ImmutableArray`1 closureDebugInfo, StateMachineTypeSymbol stateMachineTypeOpt, VariableSlotAllocator variableSlotAllocatorOpt, DebugDocumentProvider debugDocumentProvider, DiagnosticBag diagnostics, Boolean generateDebugInfo)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\Compilation\MethodCompiler.vb(961,0): at Microsoft.CodeAnalysis.VisualBasic.MethodCompiler.CompileSynthesizedMethods(TypeCompilationState compilationState)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\Compilation\MethodCompiler.vb(746,0): at Microsoft.CodeAnalysis.VisualBasic.MethodCompiler.CompileNamedType(NamedTypeSymbol symbol, Predicate`1 filter)
   F:\Enlistments\roslyn\Src\Compilers\VisualBasic\Portable\Compilation\MethodCompiler.vb(515,0): at Microsoft.CodeAnalysis.VisualBasic.MethodCompiler._Closure$__31-0._Lambda$__0()
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #2774.

@VSadov, @gafter, @agocke, @jaredpar, @khyperia Please review.

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Lowering/Rewriter.vb [hunk 1]
```diff
@@ -73,14 +75,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
                                                             compilationState,
                                                             If(symbolsCapturedWithoutCopyCtor, SpecializedCollections.EmptySet(Of Symbol)),
-                                                            diagnostics,
+                                                            localDiagnostics,
                                                             rewrittenNodes)
             End If
 
-            If bodyWithoutLambdas.HasErrors Then
+            If bodyWithoutLambdas.HasErrors OrElse localDiagnostics.HasAnyErrors Then
+                diagnostics.AddRangeAndFree(localDiagnostics)
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Lowering/Rewriter.vb]: Can hoist this statement up out of the if.

**Comment by gafter** [comment]
LGTM

File: src/Compilers/VisualBasic/Portable/Lowering/Rewriter.vb (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Lowering/Rewriter.vb]: This statement frees localDiagnostics, so we won't be able access it in the ```If``` condition.

**Comment by AlekseyTs** [comment]
- Bug number - Fixes bug #2774
- Customer scenario – Some errors reported during lowering do not result in nodes marked as bad, instead nodes in original (unlowered form) remain in the tree. Since errors are not associated with lambdas, compiler could try to emit lambdas that contained those nodes, causing code generation to throw an exception.
- Fix description - Do not run through following phases in Rewriter.LowerBodyOrInitializer once errors are encountered. This prevents, lambdas and other constructs in a method with errors from being transformed into synthetic methods and from getting to code generation.   
- Testing done - unit-tests.

@jaredpar FYI

**Comment by jaredpar** [comment]
LGTM

**Comment by MattGertz** [comment]
Approved 5/22 by ML Shiproom.

**Pull request merged**